### PR TITLE
feat: add indicator if CID is cached in local datastore

### LINKED ISSUE #8817: File Browser - Indicator of CID's cached on local node vs network
**Is your feature request related to a problem? Please describe.**

As a user of IPFS, it's difficult to visualize if files are cached on the gateway you're connected to vs if the files need to be retrieved from the network.

**Describe the solution you'd like**
An indicator to show if a CID is cached on the current gateway.  

Example (horrible) mock up:

pin

The first three entries are cached on the gateway, the final one would require pulling from other IPFS nodes.   Three indicators may be needed:

* Fully cached locally
* Partially cached locally
* Not cached locally

### PR DESCRIPTION
**Pull request by xrazis**
Closes #8817.

# Demo

A warning sign appears when the file is not cached locally. 

image

One minor issue is the 404's are printed in the console. That unfortunately cannot be avoided as it is the default behavior of the browser. Let me know if there are changes to be made code or UI wise.

### CODE REVIEW

File: assets/bindata_version_hash.go [hunk 1]
```diff
@@ -2,5 +2,5 @@
 package assets
 
 const (
-	BindataVersionHash = "512eb789cd905714e03f29d4e04de7549e8c9c3e"
+	BindataVersionHash = "d36d3139f771eb0319123dbe5df6bc82f9e5306e"
```
Comment (Jorropo) [inline assets/bindata_version_hash.go:5]: We no longer use bindata, now we use embed, pls rebase your changes on master.

File: assets/dir-index-html/src/dir-index.html [hunk 2]
```diff
@@ -19,6 +19,7 @@
...
 <link rel="stylesheet" href="style.css"/>
 <link rel="stylesheet" href="icons.css">
+<script rel="script" src="script.js" async></script>
```
Comment (Jorropo) [inline assets/dir-index-html/src/dir-index.html:22]: Why not do that on the server ? that code:
- break the noscript policy breaking `wget --recursive`.
- generate a of extra rtt and data exchange, the server already know that.

File: assets/dir-index-html/src/script.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+    const init = {
+        method: 'GET',
+        headers: {'Accept': 'application/vnd.ipfs.cache.local'}
+    };
+
+    const notCachedHTML = '<div title="File not cached locally" class="ipfs-_attention">&nbsp;</div>';
+
+    listingsPath.forEach((element, index) => {
+        fetch(element, init)
```
Comment (Jorropo) [inline assets/dir-index-html/src/script.js:16]: AFAIK this request the full file, we dont want to download GiBs just because you openned the dir listing.

**Review by Jorropo — CHANGES_REQUESTED** [review]
The idea looks good but I think that is gonna be too costy.
Files are splited in 256KiB~2MiB blocks.
It might be possible that only a few of them is there, requiring that code to check all the data.

Maybe it's not so bad if that code is server side instead, or we can do a probabilistic thing where if we found random 10 blocks we say we the file.

File: assets/dir-index-html/src/dir-index.html (hunk 2 above)
Comment (Jorropo) [reply assets/dir-index-html/src/dir-index.html:22]: Actualy this was requested for https://github.com/ipfs/go-ipfs/issues/8817#issuecomment-1075793997.
I guess the ETag make sense, but I'm not sure loosing a tiny bit of ETag is worth doing lots of extra requests.

**Comment by xrazis** [comment]
I think this is blocked for now as #8783 is pending. I will move on with requested changes once `application/vnd.ipfs.cache.local` is working.

File: assets/dir-index-html/src/script.js (hunk 3 above)
Comment (lidel) [reply assets/dir-index-html/src/script.js:16]: - Even with inexpensive `HEAD`  `application/vnd.ipfs.cache*` check we don't want to send 10k requests when opening big directory like bafybeiggvykl7skb2ndlmacg2k5modvudocffxjesexlod2pfvg5yhwrqm 
   - Even 1k `QmdmQXB2mzChmMeKY47C43LxUdg1NDJ5MWcKMKxDu7RgQm` will be wasteful, if only a small subset is visible on the screen.

This should be refactored to fetch status only for visible items – see https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API

**Review by lidel — CHANGES_REQUESTED** [review]
Some notes from go-ipfs sync today:

- This PR is unlikely to be accepted, unless we address performance concerns:
  -  simple `HEAD`, even if we have basic `application/vnd.ipfs.cache.block` won't work because dir listing already fetches root blocks of children to tell their size, so every item would be "present". 
  - only other check is to see if entire dag is in cache, and that is expensive
     - we would need `application/vnd.ipfs.cache.dag` check that is equivalent to running `ipfs dag stat --offline $CID` for every child  to see if the entire DAG is present
      - this is not something public gateways should handle, would be disabled there, but we could enable this on localhost gateway
      
Overall, a lot of complexity for a little UX indicator, I am afraid.
If you still want to work on this, we need https://github.com/ipfs/go-ipfs/issues/8783 first, then discuss performance, and then get back to this PR.

**Comment by lidel** [comment]
Closing this PR, let's continue in https://github.com/ipfs/go-ipfs/issues/8783  – design decisions need to happen there, which will inform next steps (if any).

**Pull request closed without merging**